ts/Series/Treegraph/TreegraphSeries.ts

```diff
@@ -0,0 +1,948 @@
+                series.options.link && series.options.link.lineWidth
+            ),
+            attribs = Series.prototype.pointAttribs.call(series, point, state);
+
+        if (point.isLink) {
+            attribs.stroke = linkColor;
+            attribs['stroke-width'] = linkLineWidth;
+            delete attribs.fill;
+        } else {
+            if (borderRadius) {
+                attribs.r = borderRadius;
+            }
+        }
+        return attribs;
+    }
+
+    public drawPoints(): void {
+        super.drawPoints.apply(this, arguments);
+        const points = this.points;
+        this.points = this.links as any;
+        ColumnSeries.prototype.drawPoints.apply(this, arguments);
+        this.points = points;
+    }
+    /**
+     * Run translation operations for one node.
+     * @private
+     */
+    public translateNode(point: TreegraphPoint): void {
+        const chart = this.chart,
+            node = point.node,
+            plotSizeY = chart.plotSizeY as number,
+            plotSizeX = chart.plotSizeX as number,
+            // Get the layout modifiers which are common for all nodes.
+            { ax, bx, ay, by } = this.layoutModifier,
+            x = ax * node.xPosition + bx,
+            y = ay * node.yPosition + by,
+            level = this.mapOptionsToLevel[node.level] || {},
+            markerOptions = merge(
+                this.options.marker,
+                level.marker,
+                point.options.marker
+            ),
+            symbol = markerOptions.symbol,
+            height = node.nodeSizeY,
+            width = node.nodeSizeX,
+            reversed = this.options.reversed,
+            nodeX = node.x = (chart.inverted ?
+                plotSizeX - width / 2 - x :
+                x - width / 2),
+            nodeY = node.y = (!reversed ?
+                plotSizeY - y - height / 2 :
+                y - height / 2);
+
+        point.shapeType = 'path';
+        point.plotX = nodeX;
+        point.plotY = nodeY;
+        point.shapeArgs = {
+            d: symbols[symbol || 'circle'](nodeX, nodeY, width, height),
+            x: nodeX,
+            y: nodeY,
+            width,
+            height,
+            cursor: !point.node.isLeaf ? 'pointer' : 'default'
+        };
+
+        // Set the anchor position for tooltip.
+        point.tooltipPos = chart.inverted ?
+            [plotSizeY - nodeY - height / 2, plotSizeX - nodeX - width / 2] :
+            [nodeX + width / 2, nodeY];
+    }
+}
+
+/* *
+ *
+ *  Prototype Properties
+ *
+ * */
+
+interface TreegraphSeries {
+    inverted?: boolean;
+    pointClass: typeof TreegraphPoint;
+    NodeClass: typeof TreegraphNode.Node;
+    LinkClass: typeof TreegraphLink;
+
+}
+
+namespace TreegraphSeries {
+
+}
+/* *
+ *
+ *  Registry
+ *
+ * */
+
+TreegraphSeries.prototype.pointClass = TreegraphPoint;
+TreegraphSeries.prototype.NodeClass = TreegraphNode.Node;
+TreegraphSeries.prototype.LinkClass = TreegraphLink;
```

Comment:
- empty namespace can be removed
- property assignment should come before Registry section
- do we need the other point classes really on the prototype? The pointClass is for sake of compatibility.

Reply:
I added an extra reference to `NodeClass` in a similar fashion to `pointClass` because I figured, that it would be a way of creating the object using the correct class in Treemap/Sunburst/Treegraph series. When overwriting this property, the method `buildTree` creates correct nodes: https://github.com/highcharts/highcharts/blob/c994a3a6d2500f0473b1aa6c96bc892c8a6a05d1/ts/Series/Treemap/TreemapSeries.ts#L1294-L1302

Is there a better way?

Reply:
It is possible to move it to the constructor, but I am not happy with the required type hack. So I have no objection anymore. Just for curiosity the necessary changes would look as follows:

```diff
diff --git a/ts/Series/Sunburst/SunburstSeries.ts b/ts/Series/Sunburst/SunburstSeries.ts
index 469330732f..28130f42f3 100644
--- a/ts/Series/Sunburst/SunburstSeries.ts
+++ b/ts/Series/Sunburst/SunburstSeries.ts
@@ -418,6 +418,8 @@ class SunburstSeries extends TreemapSeries {
      *
      * */
 
+    public static readonly NodeClass = SunburstNode;
+
     /**
      * A Sunburst displays hierarchical data, where a level in the hierarchy is
      * represented by a circle. The center represents the root node of the tree.
@@ -1153,10 +1155,10 @@ class SunburstSeries extends TreemapSeries {
  * */
 
 interface SunburstSeries {
+    constructor: typeof SunburstSeries;
     getCenter: typeof CU['getCenter'];
     pointClass: typeof SunburstPoint;
     utils: typeof SunburstUtilities;
-    NodeClass: typeof SunburstNode;
 }
 extend(SunburstSeries.prototype, {
     drawDataLabels: noop, // drawDataLabels is called in drawPoints
@@ -1165,9 +1167,8 @@ extend(SunburstSeries.prototype, {
     onPointSupported: true,
     pointAttribs: ColumnSeries.prototype.pointAttribs as any,
     pointClass: SunburstPoint,
-    NodeClass: SunburstNode,
     utils: SunburstUtilities
-});
+} as SunburstSeries);
 
 /* *
  *
diff --git a/ts/Series/Treegraph/TreegraphNode.ts b/ts/Series/Treegraph/TreegraphNode.ts
index 65f733fb0b..c734dc8aac 100644
--- a/ts/Series/Treegraph/TreegraphNode.ts
+++ b/ts/Series/Treegraph/TreegraphNode.ts
@@ -25,9 +25,7 @@ import SeriesRegistry from '../../Core/Series/SeriesRegistry.js';
 const {
     seriesTypes: {
         treemap: {
-            prototype: {
-                NodeClass: TreemapNode
-            }
+            NodeClass: TreemapNode
         }
     }
 } = SeriesRegistry;
diff --git a/ts/Series/Treegraph/TreegraphSeries.ts b/ts/Series/Treegraph/TreegraphSeries.ts
index 87066113be..4976afbd4e 100644
--- a/ts/Series/Treegraph/TreegraphSeries.ts
+++ b/ts/Series/Treegraph/TreegraphSeries.ts
@@ -69,12 +69,17 @@ interface LayoutModifiers {
  * @augments Highcharts.Series
  */
 class TreegraphSeries extends TreemapSeries {
+
     /* *
      *
      *  Static Properties
      *
      * */
 
+    public static readonly LinkClass = TreegraphLink;
+
+    public static readonly NodeClass = TreegraphNode;
+
     /**
      * A treegraph series is a diagram, which shows a relation between ancestors
      * and descendants with a clear parent - child relation.
@@ -419,7 +424,7 @@ class TreegraphSeries extends TreemapSeries {
             if (point.node.parent) {
                 const pointOptions = merge(levelOptions, point.options);
                 if (!point.linkToParent || !point.linkToParent.update) {
-                    const link = new series.LinkClass().init(
+                    const link = new series.constructor.LinkClass().init(
                         series,
                         pointOptions,
                         void 0,
@@ -808,16 +813,11 @@ class TreegraphSeries extends TreemapSeries {
  * */
 
 interface TreegraphSeries {
+    constructor: typeof TreegraphSeries;
     inverted?: boolean;
     pointClass: typeof TreegraphPoint;
-    NodeClass: typeof TreegraphNode;
-    LinkClass: typeof TreegraphLink;
-
 }
 
-namespace TreegraphSeries {
-
-}
 /* *
  *
  *  Registry
@@ -825,8 +825,6 @@ namespace TreegraphSeries {
  * */
 
 TreegraphSeries.prototype.pointClass = TreegraphPoint;
-TreegraphSeries.prototype.NodeClass = TreegraphNode;
-TreegraphSeries.prototype.LinkClass = TreegraphLink;
 declare module '../../Core/Series/SeriesType' {
     interface SeriesTypeRegistry {
         treegraph: typeof TreegraphSeries;
diff --git a/ts/Series/Treemap/TreemapSeries.ts b/ts/Series/Treemap/TreemapSeries.ts
index 85c6080a05..29003bd506 100644
--- a/ts/Series/Treemap/TreemapSeries.ts
+++ b/ts/Series/Treemap/TreemapSeries.ts
@@ -100,6 +100,8 @@ class TreemapSeries extends ScatterSeries {
      *
      * */
 
+    public static readonly NodeClass = TreemapNode;
+
     /**
      * A treemap displays hierarchical data using nested rectangles. The data
      * can be laid out in varying ways depending on options.
@@ -1291,7 +1293,7 @@ class TreemapSeries extends ScatterSeries {
             height = Math.max(child.height + 1, height);
             children.push(child);
         });
-        node = new series.NodeClass().init(
+        node = new series.constructor.NodeClass().init(
             id,
             index,
             children,
@@ -1922,6 +1924,7 @@ class TreemapSeries extends ScatterSeries {
 interface TreemapSeries extends ColorMapComposition.SeriesComposition, TU.Series {
     colorAttribs: ColorMapComposition.SeriesComposition['colorAttribs'];
     colorKey: string;
+    constructor: typeof TreemapSeries;
     directTouch: boolean;
     drawLegendSymbol: typeof LegendSymbol.drawRectangle;
     getExtremesFromAll: boolean;
@@ -1929,7 +1932,6 @@ interface TreemapSeries extends ColorMapComposition.SeriesComposition, TU.Series
     parallelArrays: Array<string>;
     pointArrayMap: Array<string>;
     pointClass: typeof TreemapPoint;
-    NodeClass: typeof TreemapNode;
     trackerGroups: Array<string>;
     utils: {
         recursive: typeof TreemapUtilities.recursive;
@@ -1947,12 +1949,11 @@ extend(TreemapSeries.prototype, {
     parallelArrays: ['x', 'y', 'value', 'colorValue'],
     pointArrayMap: ['value'],
     pointClass: TreemapPoint,
-    NodeClass: TreemapNode,
     trackerGroups: ['group', 'dataLabelsGroup'],
     utils: {
         recursive: TreemapUtilities.recursive
     }
-});
+} as TreemapSeries);
 ColorMapComposition.compose(TreemapSeries);
 
 /* *
```